plugins/node/opentelemetry-instrumentation-undici/src/undici.ts

```diff
@@ -0,0 +1,460 @@
+  }
+
+  // This is the 3rd message we get for each request and it's fired when the server
+  // headers are received, body may not be accessible yet.
+  // From the response headers we can set the status and content length
+  private onResponseHeaders({
+    request,
+    response,
+  }: ResponseHeadersMessage): void {
+    const record = this._recordFromReq.get(request);
+
+    if (!record) {
+      return;
+    }
+
+    const { span, attributes, startTime } = record;
+    // We are currently *not* capturing response headers, even though the
+    // intake API does allow it, because none of the other `setHttpContext`
+    // uses currently do
+    const spanAttributes: Attributes = {
+      [SemanticAttributes.HTTP_RESPONSE_STATUS_CODE]: response.statusCode,
+    };
+
+    // Get headers with names lowercased but values intact
+    const resHeaders = new Map<string, string>();
+    for (let idx = 0; idx < response.headers.length; idx = idx + 2) {
+      resHeaders.set(
+        response.headers[idx].toString().toLowerCase(),
+        response.headers[idx + 1].toString()
+      );
+    }
+
+    // Put response headers as attributes based on config
+    const config = this._getConfig();
+    if (config.headersToSpanAttributes?.responseHeaders) {
+      config.headersToSpanAttributes.responseHeaders
+        .map(name => name.toLowerCase())
+        .filter(name => resHeaders.has(name))
+        .forEach(name => {
+          spanAttributes[`http.response.header.${name}`] = resHeaders.get(name);
+        });
+    }
+
+    // `content-length` header is a special case
+    if (resHeaders.has('content-length')) {
+      const contentLength = Number(resHeaders.get('content-length'));
+      if (!isNaN(contentLength)) {
+        spanAttributes['http.response.header.content-length'] = contentLength;
+      }
+    }
+
+    span.setAttributes(spanAttributes);
+    span.setStatus({
+      code:
+        response.statusCode >= 400
+          ? SpanStatusCode.ERROR
+          : SpanStatusCode.UNSET,
+    });
+    this._recordFromReq.set(request, {
+      span,
+      startTime,
+      attributes: Object.assign(attributes, spanAttributes),
+    });
+  }
+
+  // This is the last event we receive if the request went without any errors
+  private onDone({ request, response }: RequestTrailersMessage): void {
+    const record = this._recordFromReq.get(request);
+
+    if (!record) {
+      return;
+    }
+
+    const config = this._getConfig();
+    const { span, attributes, startTime } = record;
+
+    // Let the user apply custom attribs before ending the span
+    safeExecuteInTheMiddle(
+      () => config.applyCustomAttributesOnSpan?.(span, request, response),
+      e =>
+        e && this._diag.error('caught applyCustomAttributesOnSpan error: ', e),
+      true
+    );
+
+    // End the span
+    span.end();
+    this._recordFromReq.delete(request);
+
+    // Record metrics
+    this.recordRequestDuration(attributes, startTime);
```

Comment:
The *.NET*-specific semconv docs for this metric (https://opentelemetry.io/docs/specs/semconv/dotnet/dotnet-http-metrics/#metric-httpclientrequestduration) state:

> Client request duration measures the time it takes to receive response headers and doesn’t include time to read the response body.

If we were to follow the same, we would want to record the request duration in `onResponseHeaders`. I can see .NET's point of view, because the response time is not skewed by large response bodies. However, I don't know what other languages do here. I haven't dug into the language-agnostic OTel spec/semconv to see if it provides guidance.

Reply:
https://opentelemetry.io/docs/specs/semconv/http/http-spans/#http-client-span-duration

> There are some minimal constraints that SHOULD be honored:
> - HTTP client spans SHOULD start sometime before the first request byte is sent. This may or may not include connection time.
> - HTTP client spans SHOULD end sometime after the HTTP response headers are fully read (or when they fail to be read). This may or may not include reading the response body.
>
>If there is any possibility for application code to not fully read the HTTP response (and for the HTTP client library to then have to clean up the HTTP response asynchronously), the HTTP client span SHOULD NOT be ended in this cleanup phase, and instead SHOULD end at some point after the HTTP response headers are fully read (or fail to be read). This avoids the span being ended asynchronously later on at a time which is no longer directly associated with the application code which made the HTTP request.
>
> Because of the potential for confusion around this, HTTP client library instrumentations SHOULD document their behavior around ending HTTP client spans.

If I understand correctly the case for this instrumentation would be that we should end the span here, but we should document that behavior.

Reply:
@pichlermarc do you mean to add it to the README?

Reply:
@pichlermarc @trentm 

I've added a new section in [this commit](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1951/commits/a9ad5118bc7e0b24e670126e3f651cb57c41da0a). Let me know what you think